src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,53 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static final int GROWTH_FACTOR = 2;
+    private T[] stack;
+    private int size;
+    private int capacity;
 
+    public MateStack() {
+        final int default_capacity = 10;
+        stack = (T[]) new Object[default_capacity];
+        size = 0;
+        capacity = default_capacity;
     }
 
-    public T peek() {
-        return null;
+    public void push(T value) {
+        if (size >= capacity) {
+            grow();
+        }
+        stack[size] = value;
+        size++;
     }
 
-    public T pop() {
-        return null;
+    public T peek() throws EmptyStackException {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        return stack[size - 1];
+    }
+
+    public T pop() throws EmptyStackException {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        T poppedElement = stack[size - 1];
+        stack[size - 1] = null;
+        size--;
+        return poppedElement;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void grow() {
+        T[] oldStack = stack;
+        capacity *= GROWTH_FACTOR;
+        stack = (T[]) new Object[capacity];
+        System.arraycopy(oldStack, 0, stack, 0, size);
```

Comment:
The `grow` method is private and correctly resizes the internal array when the stack reaches its capacity. Consider checking for integer overflow when increasing the capacity, as multiplying by the growth factor could exceed `Integer.MAX_VALUE`.